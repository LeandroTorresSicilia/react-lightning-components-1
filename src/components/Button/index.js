/* eslint-disable react/require-default-props */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export default class Button extends Component {
    getBtnClass() {
        const { variant, className } = this.props;

        return classnames('slds-button', {
            'slds-button--neutral': variant === 'neutral',
            'slds-button--brand': variant === 'brand',
            'slds-button--inverse': variant === 'inverse',
            'slds-button--destructive': variant === 'destructive',
            'slds-button--success': variant === 'success',
        }, className);
    }

    render() {
        const {
            label,
            disabled,
            tabIndex,
            onFocus,
            onBlur,
            onClick,
            style,
            children,
            title,
            ariaHaspopup,
        } = this.props;

        return (
            <button className={this.getBtnClass()}
                    disabled={disabled}
                    tabIndex={tabIndex}
                    onFocus={onFocus}
                    onBlur={onBlur}
                    onClick={onClick}
                    title={title}
                    style={style}
                    aria-haspopup={ariaHaspopup} >
                { !children ? label : children }
            </button>
        );
    }
}

Button.propTypes = {
    /** Label for the button. If a children is specified, it will be displayed.
     Otherwise, the component will expect a label which will then displayed */
    label: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.node,
    ]),
    /** Indicates that the element has a popup context menu or sub-level menu. It is
     used for srceen readers */
    ariaHaspopup: PropTypes.bool,
    /** A button predefined style. It must be a string */
    variant: PropTypes.oneOf(['neutral', 'brand', 'inverse', 'destructive', 'success', 'bare']),
    /** This is what will be displayed inside the button. It has precedence over
     the label prop */
    children: PropTypes.node,
    /** Class for custom styles */
    className: PropTypes.string,
    /** Disables the button if set to true */
    disabled: PropTypes.bool,
    /** Tab index */
    tabIndex: PropTypes.number,
    /** Callback function fired when the button is clicked */
    onClick: PropTypes.func,
    /** Callback function fired when the button is focused */
    onFocus: PropTypes.func,
    /** Callback function fired when the button is blurred */
    onBlur: PropTypes.func,
    /** Object with the custom styles. The properties must be
     in camelCase naming convention (e.g. { fontFamily: ‘helvetica’ }) */
    style: PropTypes.object,
    /** This is a description that is showed when a user hover the button */
    title: PropTypes.string,
};

Button.defaultProps = {
    ariaHaspopup: false,
    variant: 'neutral',
    disabled: false,
    onClick: () => {},
    onFocus: () => {},
    onBlur: () => {},
};
