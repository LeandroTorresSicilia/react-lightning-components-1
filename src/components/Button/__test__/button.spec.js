import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Button from './../index';

describe('<Button/>', () => {
    it('should have proper classes when variant is brand', () => {
        const component = renderer.create(<Button label="Button label" variant="brand" />);

        expect(component).toMatchSnapshot();
    });

    it('should call onClick function when someone click over', () => {
        const onClickMockFn = jest.fn();
        const component = shallow(<Button label="Button label" onClick={onClickMockFn} />);

        component.simulate('click');
        expect(onClickMockFn.mock.calls.length).toBe(1);
    });

    it('should call onBlur function when it lost the focus', () => {
        const onBlurMockFn = jest.fn();
        const component = shallow(<Button label="Button label" onBlur={onBlurMockFn} />);

        component.simulate('blur');
        expect(onBlurMockFn.mock.calls.length).toBe(1);
    });

    it('should call onFocus function when it gets the focus', () => {
        const onFocusMockFn = jest.fn();
        const component = shallow(<Button label="Button label" onFocus={onFocusMockFn} />);

        component.simulate('focus');
        expect(onFocusMockFn.mock.calls.length).toBe(1);
    });

    it('should render a children node', () => {
        const component = shallow(<Button label={<p>Button label</p>} ><p>Child node</p></Button>);

        expect(component.contains(<p>Child node</p>)).toBe(true);
    });

    it('should have not <p>Button label</p>', () => {
        const component = shallow(
            <Button label={<p className="button-label">Button label</p>} >
                <p>Child node</p>
            </Button>);

        expect(component.find('.button-label').exists()).toBe(false);
    });

    it('should have only one children', () => {
        const component = shallow(<Button label={<p>Button label</p>} ><p>Child node</p></Button>);

        expect(component.children()).toHaveLength(1);
    });

    it('should have the Child node text', () => {
        const component = shallow(<Button label={<p>Button label</p>} ><p>Child node</p></Button>);

        expect(component.text()).toBe('Child node');
    });
});
