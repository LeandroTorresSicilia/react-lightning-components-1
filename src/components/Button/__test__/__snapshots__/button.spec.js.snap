// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should have proper classes when variant is brand 1`] = `
<button
  aria-haspopup={false}
  className="slds-button slds-button--brand"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={undefined}
  tabIndex={undefined}
  title={undefined}
>
  Button label
</button>
`;
